use `std::vector<DetId>` as data format of SiStrip `DetId`s with FED errors

missirol opened the pull request:
#### PR description:

This PR changes the data format of the collection used for the DetIds of SiStrip modules with FED errors (this collection is used to mask said modules during the event reconstruction).

#42475 showed that this information needs to be added to the RawPrime event content (see that PR for details).

The current data format for this is `DetIdCollection`, and concerns were raised (see https://github.com/cms-sw/cmssw/pull/42495#issuecomment-1674956299 and replies) on the forward-compatibility of this data format in view of possibly adding it to the RAW data tier (which is expected to be backward- and forward-compatible across release cycles).

Following the discussion in https://github.com/cms-sw/cmssw/pull/42495#issuecomment-1674956299, this PR tries to implement a solution replacing `DetIdCollection` (== `edm::EDCollection<DetId>`) with `DetIdVector` (== `std::vector<DetId>`) in the SiStrip unpacker and related clients/utilities, such that the data format added to RAW(Prime) would be a `std::vector<DetId>` rather than a `edm::EDCollection<DetId>`.

Merely technical. No changes expected.

Thanks to @mmusich for help with preparing this PR.

#### PR validation:

Not much. Tested `addOnTests` on MC, and a few `runTheMatrix` wfs.

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

`CMSSW_13_2_X`

Related to the 2023 HIon data-taking.

missirol committed:
use std::vector<DetId> as data format of SiStrip DetIds with FED errors

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42662/36725

- This PR adds an extra 144KB to repository

cmsbuild commented:
A new Pull Request was created by  @missirol (Marino Missiroli) for master.

It involves the following packages:

- CalibTracker/SiStripHitEfficiency (**alca**)
- Calibration/TkAlCaRecoProducers (**alca**)
- Configuration/EventContent (**operations**)
- Configuration/Skimming (**pdmv**)
- DPGAnalysis/SiStripTools (**analysis**)
- DPGAnalysis/Skims (**pdmv**)
- DQM/SiPixelHeterogeneous (**dqm**)
- DataFormats/DetId (**simulation**)
- EventFilter/SiStripRawToDigi (**reconstruction**)
- RecoLocalTracker/Configuration (**reconstruction**)
- RecoTracker/MeasurementDet (**reconstruction**)
- SimG4Core/Configuration (**simulation**)


@rappoccio, @bbilin, @pmandrik, @tjavaid, @perrotta, @civanch, @cmsbuild, @miquork, @antoniovilela, @mandrenguyen, @davidlange6, @consuegs, @emanueleusai, @mdhildreth, @AdrianoDee, @syuvivida, @micsucmed, @fabiocos, @francescobrivio, @nothingface0, @clacaputo, @saumyaphor4252, @sunilUIET, @rvenditti, @tvami can you please review it and eventually sign? Thanks.
@echabert, @VourMa, @felicepantaleo, @yduhm, @robervalwalsh, @Martin-Grunewald, @bsunanda, @fioriNTU, @threus, @mmusich, @slomeo, @venturia, @youyingli, @makortel, @JanFSchulte, @dgulhan, @jlidrych, @GiacomoSguazzoni, @rovere, @VinInn, @tocheng, @jandrea, @alesaggio, @idebruyn, @ebrondol, @mtosi, @fabiocos, @AnnikaStein, @gbenelli, @gpetruc this is something you requested to watch as well.
@perrotta, @dpiparo, @antoniovilela, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

fwyzard commented:
what about just changing the definition of `DetIdCollection` from
```c++
typedef edm::EDCollection<DetId> DetIdCollection;
```
to
```c++
using DetIdCollection = std::vector<DetId>;
```
?

Would that be more or less intrusive?

mmusich commented:
> Would that be more or less intrusive?

certainly cheaper in term of changeset, but what if we want to go back to re-analyze the old persisted data (e.g. in ALCARECO samples) taken in < 13.2.X with a new release? With this solution (if we need it) we could at least plug  some sort of converter upfront.

fwyzard commented:
> what if we want to go back to re-analyze the old persisted data (e.g. in ALCARECO samples) taken in < 13.2.X with a new release? With this solution (if we need it) we could at least plug some sort of converter upfront.

Good point, thanks.

fwyzard commented:
please test

mandrenguyen commented:
urgent

Needed for PbPb data taking

fwyzard commented:
These changes break compatibility with old AlCaRaw data.

This is expected, but it affects some of the unit tests: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17e62e/34486/unitTests/src/CalibTracker/SiStripHitEfficiency/test/testSiStripHitEfficiency/testing.log .

How do people want to proceed ?

mmusich commented:
> How do people want to proceed ?

Ideally we should detect at rutime which format we have to deal with and act consequently. If we are short in time we can silence the tests for now while we figure how to support the old data.

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17e62e/34486/summary.html
**COMMIT**: 2cff7710f362f08fb00d072fe11a8cf390c721b2
**CMSSW**: CMSSW_13_3_X_2023-08-24-2300/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42662/34486/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found errors in the following unit tests:

<pre>---> test testSiStripHitEfficiency had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially added 2 lines to the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 10 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3153095
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3153067
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-08-24-2300+17e62e/58605/triggerResults): no differences found

rappoccio commented:
please test

* Try to refresh tests to see if they pass.

mmusich commented:
please abort
  * they will not pass for sure, see https://github.com/cms-sw/cmssw/pull/42662#issuecomment-1693270042

makortel commented:
For what its worth, from `core` perspective this PR looks good towards ensuring backward compatibility (and also helps towards forward compatibility, but that is a whole different topic and outside of the scope of this PR).

missirol commented:
df93d7d39e2 adds a workaround for the unit test. Feel free to suggest improvements / alternatives.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42662/36732

- This PR adds an extra 140KB to repository

cmsbuild commented:
Pull request #42662 was updated. @rappoccio, @bbilin, @pmandrik, @tjavaid, @perrotta, @civanch, @cmsbuild, @miquork, @antoniovilela, @mandrenguyen, @davidlange6, @consuegs, @emanueleusai, @mdhildreth, @AdrianoDee, @syuvivida, @micsucmed, @fabiocos, @francescobrivio, @nothingface0, @clacaputo, @saumyaphor4252, @sunilUIET, @rvenditti, @tvami can you please check and sign again.

mmusich commented:
test parameters:
   * workflows = 1001.2, 1001.3

mmusich commented:
please test

slava77 commented:
@makortel @cms-sw/core-l2 
what's the actual problem with the `EDCollection<DetId>`? Is it the fact that it's a template or that it's a class containing a vector?
The PR here switches to `vector<DetId>`.

FEDRawData is a class containing a vector of chars and is supported.

If a `vector<DetId>` is OK, I would think that a class containing such a vector should also be OK. So, what remains is the template declaration.

mmusich commented on CalibTracker/SiStripHitEfficiency/plugins/SiStripHitEfficiencyWorker.cc:
```diff
@@ -445,11 +449,31 @@ void SiStripHitEfficiencyWorker::analyze(const edm::Event& e, const edm::EventSe
+  auto const& fedErrorIdsCol_h = e.getHandle(digisCol_token_);
+  auto const& fedErrorIdsVec_h = e.getHandle(digisVec_token_);
+  DetIdVector fedErrorIdsVecFromColl;
+  if (not fedErrorIdsVec_h.isValid()) {
```
out of curiosity I have tried myself a similar approach but using `failedToGet` instead of `isValid`, what's the actual difference between the two? maybe some core expert could educate me on that (apologies I could not find in the manual an answer).

fwyzard replied:
According to a comment in `DataFormats/Common/interface/Handle.h`:

> To check validity, one can use the `isValid()` function.
> 
> If `failedToGet()` returns true then the requested data is not available
> If `failedToGet()` returns false but `isValid()` is also false then no attempt 
>   to get data has occurred

fwyzard replied:
(assuming it is still accurate)

missirol commented:
Naive question: do we need, or want, to apply a workaround like df93d7d39e2 beyond the failing unit tests ?

makortel commented:
> what's the actual problem with the `EDCollection<DetId>`? Is it the fact that it's a template or that it's a class containing a vector?

The `edm::EDCollection<T>` is a remnant from long time ago, and replicates (unnecessarily) the functionality of `std::vector`. It would be nice to clean it up at some point (in favor of `std::vector`). If the `EDCollection<T>` would become part of RAW, we could never remove it.

fwyzard commented:
> Naive question: do we need, or want, to apply a workaround like [df93d7d](https://github.com/cms-sw/cmssw/commit/df93d7d39e21dc40df64b7b503da7634f76990f5) beyond the failing unit tests ? (e.g. in plugins used offline for reconstruction, or some other purpose)

I would suggest a better workaround:
  - add to to `EDCollection` the method
    ```c++
    std::vector<T> const& as_vector() const { return obj; }
    ```
  - simplify the workaround to avoid the copy, and use
    ```c++
    auto const& fedErrorIds = fedErrorIdsVec_h.isValid() ? *fedErrorIdsVec_h : fedErrorIdsCol_h->as_vector();
    ```

missirol committed:
add as_vector method to edm::EDCollection<T> class template

Suggested by @fwyzard in
https://github.com/cms-sw/cmssw/pull/42662#issuecomment-1693530380

missirol committed:
workaround for unit test "testSiStripHitEfficiency"

Suggested by @fwyzard in
https://github.com/cms-sw/cmssw/pull/42662#issuecomment-1693530380

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-42662/36734

- This PR adds an extra 48KB to repository

cmsbuild commented:
Pull request #42662 was updated. @rappoccio, @bbilin, @pmandrik, @tjavaid, @perrotta, @civanch, @makortel, @cmsbuild, @miquork, @antoniovilela, @mandrenguyen, @davidlange6, @smuzaffar, @consuegs, @Dr15Jones, @emanueleusai, @mdhildreth, @AdrianoDee, @syuvivida, @micsucmed, @fabiocos, @francescobrivio, @nothingface0, @clacaputo, @saumyaphor4252, @sunilUIET, @tvami, @rvenditti can you please check and sign again.

mmusich commented:
please test

sextonkennedy commented:
Glad to see this has progressed today.  I did exchange email with Matti today as promised at the XC.  He said, "the PR looks good from core point of view."

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-17e62e/34491/summary.html
**COMMIT**: acbbf855391d6e133e8172ca5111e2d27bcce1c3
**CMSSW**: CMSSW_13_3_X_2023-08-25-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/42662/34491/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially removed 6 lines from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 1 differences found in the comparisons
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 3168030
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3168008
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 218 log files, 161 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_3_X_2023-08-25-1100+17e62e/58610/triggerResults): no differences found

makortel commented:
+core

mandrenguyen commented:
+reconstruction

emanueleusai commented:
+1

perrotta commented:
+alca

perrotta commented:
@cms-sw/simulation-l2 @cms-sw/pdmv-l2 please have a look at this PR

perrotta commented:
+operations

civanch commented:
+1

perrotta commented:
+1
- @cms-sw/pdmv-l2 let merge this in the master, so that it can be tested in the IBs. Please comment (and sign, if you think so) on the backported PR

perrotta commented:
merge
